File: scripts/rlock/rlock.py
#!/usr/bin/env python3

import time
import threading

rlock = threading.RLock()

def worker(id):
    with rlock:
        print(f"Thread: {id} acquired the lock.")
        with rlock:
            print(f"Thread: {id} re-acquired the same lock.")
            time.sleep(1)
        print(f"Thread: {id} released the nested lock.")
    print(f"Thread: {id} released the lock.")

if __name__ == '__main__':
    threading.Thread(target=worker, args=(1,)).start()

File: scripts/rlock/rlock.pl
#!/usr/bin/env perl

use v5.38;
use threads;
use threads::shared;

my $rlock :shared = 1;

sub worker {
    my $id = threads->tid();
    lock($rlock);
    say "Thread: $id acquired the lock.";
    {
        say "Thread: $id re-acquired the same lock.";
        lock($rlock);
        sleep(1);
        say "Thread: $id released the nested lock.";
    }
    say "Thread: $id released the lock.";
}

threads->create(\&worker)->join();
